rectools/visuals/metrics_app.py

```diff
@@ -149,61 +258,106 @@ def _update_chart(
         metric_y: widgets.Dropdown,
         use_avg: widgets.Checkbox,
         fold_i: widgets.Dropdown,
-    ) -> None:
-        data = self._make_chart_data_avg() if use_avg.value else self._make_chart_data(fold_i.value)
-        existing_traces = {trace.name: trace for trace in fig_widget.data}
+        meta_feature: widgets.Dropdown,
+        use_meta: widgets.Checkbox,
+    ) -> None:  # pragma: no cover
+        chart_data = self._create_chart_data(use_avg, fold_i)
+        color_clmn = meta_feature.value if use_meta.value else Columns.Model
+
+        # Save dots symbols from the previous widget state
+        # `split(" ", 1)[-1]` removed metainfo from trace name. Thus we guarantee to map with traces from previous state
+        trace_name2symbol = {trace.name.split(" ", 1)[-1]: trace.marker.symbol for trace in self.fig.data}
+        legend_title = f"{meta_feature.value}, {DEFAULT_LEGEND_TITLE}" if use_meta.value else DEFAULT_LEGEND_TITLE
+        self.fig = self._create_chart(chart_data, metric_x.value, metric_y.value, color_clmn, legend_title)
+
+        for trace in self.fig.data:
+            trace_name = trace.name.split(" ", 1)[-1]
+            trace.marker.symbol = trace_name2symbol[trace_name]
+
+        nometa_trace_name2idx = {trace.name.split(" ", 1)[-1]: idx for idx, trace in enumerate(self.fig.data)}
+
         with fig_widget.batch_update():
-            for model in self.model_names:
-                df = data[data[Columns.Model] == model]
-                hover = f"<b>{model}</b><br>{metric_x.value}: %{{x}}<br>{metric_y.value}: %{{y}}<extra></extra>"
-                if model in existing_traces:  # pragma: no cover
-                    trace = existing_traces[model]
-                    trace.x = df[metric_x.value]
-                    trace.y = df[metric_y.value]
-                    trace.hovertemplate = hover
-                else:
-                    fig_widget.add_scatter(
-                        x=df[metric_x.value],
-                        y=df[metric_y.value],
-                        mode="markers",
-                        name=model,
-                        hovertemplate=hover,
-                        showlegend=self.show_legend,
-                    )
-            fig_widget.layout.xaxis.title = metric_x.value
-            fig_widget.layout.yaxis.title = metric_y.value
-        self.fig = go.Figure(data=fig_widget.data, layout=fig_widget.layout)
-        self.fig.layout.margin = None
+            for trace in self.fig.data:
```

Comment:
Please double check me. But I think that you can just do enumerate here. Because you were enumerating `self.fig.data` to create `nometa_trace_name2idx`. And now you are iterating through the same self.fig.data. We can drop `nometa_trace_name2idx` and just enumerate here to get `idx`.

Or is there something that I am missing? Then please write a comment for the logic. 
I'm not familiar with plotly figure widgets

Reply:
No, unfortunately, it is more complicated.
The chart changes trace orders for "meta" and "no-meta" modes, which is evident in the legend sorting section. Therefore, I cannot simply iterate over them. Additionally, the meta parameter is now part of the trace name, necessitating its removal to match trace names between the "meta" and "no-meta" states. As a result, I have implemented this key mapping approach.

Reply:
Updated